Docs: Add article on how to implement runtime checks (#1468)
grafana/plugin-tools: docusaurus/docs/how-to-guides/runtime-checks.md

---
id: runtime-checks
title: Manage backwards compatibility with runtime checks
description: How to manage backwards compatibility with runtime checks.
keywords:
  - grafana
  - plugins
  - plugin
  - compatibility
---

> # Manage backwards compatibility with runtime checks

Author: Good point! Rewrote the first paragraph to better describe this.

> To leverage new Grafana features in plugins while ensuring backward compatibility with older versions, plugin authors must implement conditional logic to adapt to feature availability at runtime. Neglecting backward compatibility can lead to plugin crashes and a degraded user experience.

Reviewer: ```suggestion
To take advantage of new Grafana features in plugins while maintaining compatibility with older versions, plugin authors need to implement conditional logic that checks for feature availability during runtime. Failing to account for backward compatibility can result in plugin crashes and a poor user experience.
```

> The approach for performing these runtime checks depends on the specific feature and how it is exposed to plugin developers. Below are examples illustrating how to handle such scenarios effectively.

Reviewer: ```suggestion
The method for performing these runtime checks varies depending on the feature and how it is made available to plugin developers. The following examples demonstrate best practices for handling these scenarios effectively.
```

> ## Example: Conditionally invoking functions

Reviewer: ```suggestion
## Example: Conditionally invoke functions
```

> ## Example: Conditionally invoking functions

Reviewer: ```suggestion
## Example: Conditionally invoke functions
```

> Starting with Grafana 10.1.0, the `@grafana/data` package introduced the `createDataFrame` function, marking the deprecation of the `MutableDataFrame` class. To maintain compatibility with versions prior to 10.1.0, plugins must implement conditional logic to determine whether these APIs are available at runtime.

Reviewer: ```suggestion
Grafana 10.1.0 introduced the `createDataFrame` function in the `@grafana/data` package, deprecatiing the `MutableDataFrame` class. To maintain compatibility with Grafana versions prior to 10.1.0, plugins must implement conditional logic to determine whether these APIs are available at runtime.
```

```tsx
import { createDataFrame, DataFrameDTO, MutableDataFrame } from '@grafana/data';

function getDataFrame(data: DataFrameDTO) {
  if (typeof createDataFrame !== 'undefined') {
    // use the new API if available
    return createDataFrame(data);
  } else {
    // fallback to the deprecated class for older versions
    return new MutableDataFrame(data);
  }
}
```

> ## Example: Conditionally using React hooks

Reviewer: ```suggestion
## Example: Conditionally use React hooks
```

> In Grafana 11.1.0, the syncronous `getPluginLinkExtensions` function got depcrecated in favour of the new reactive `usePluginLinks` hook. The following example shows how to dynamically alternate between the two APIs based on their availability.

Reviewer: ```suggestion
In Grafana 11.1.0, the synchronous `getPluginLinkExtensions` function was deprecated and replaced by the reactive `usePluginLinks` hook. The following example demonstrates how to dynamically switch between the two APIs based on their availability.
```

```tsx
import { useMemo } from 'react';
import { PluginExtensionLink } from '@grafana/data';
import {
  GetPluginExtensionsOptions,
  getPluginLinkExtensions,
  usePluginLinks as usePluginLinksOriginal,
} from '@grafana/runtime';

function useLegacyLinkExtensions({ context, extensionPointId }: GetPluginExtensionsOptions): {
  links: PluginExtensionLink[];
  isLoading: boolean;
} {
  const { extensions } = useMemo(
    () =>
      getPluginLinkExtensions({
        extensionPointId,
        context,
      }),
    [context, extensionPointId]
  );

  return {
    links: extensions,
    isLoading: false,
  };
}

// dynamically decide which API to use
const usePluginLinks = usePluginLinksOriginal !== undefined ? usePluginLinksOriginal : useLegacyLinkExtensions;

export function ToolbarExtensionPoint() {
  const { links, isLoading } = usePluginLinks({ extensionPointId: 'myorg-foo-app/toolbar/v1' });

   // Your implementation here
  ...
}
```

> ## Example: Conditionally rendering a React component

Reviewer: ```suggestion
## Example: Conditionally render a React component
```

> The `UserIcon` was introduced in Grafana 10.1.0 and does not have a predecessor in earlier versions. To ensure compatibility, you can render the UserIcon component only when it is available in the current runtime environment.

Reviewer: ```suggestion
The `UserIcon` component was introduced in Grafana 10.1.0 and has no equivalent in earlier versions. To maintain compatibility, render the UserIcon component only if it is available in the current runtime environment.
```

```tsx
import React from 'react';
import { Card, UserIcon, UserView } from '@grafana/ui';

export const Profile = ({ userView }: { userView: UserView }) => {
  return (
    <Card>
      <Card.Heading>Profile</Card.Heading>
      <Card.Meta>{['Tag 1']}</Card.Meta>
      {/* Conditionally render the UserIcon component if it exists */}
      {UserIcon && <UserIcon userView={userView} />}
    </Card>
  );
};
```

## Example: Cover conditional rendering in an end-to-end test

When a feature is only available in certain Grafana versions, it’s a good practice to validate its conditional rendering through end-to-end (E2E) tests. These tests ensure that the plugin behaves correctly in both newer environments where the feature exists and older environments where it is unavailable.

> In the following example, the test verifies that the `UserIcon` is rendered only if the Grafana version is 10.1.0 or later, while the rest of the user profile is always rendered.

Reviewer: ```suggestion
The following example tests whether the `UserIcon` component is rendered only in Grafana 10.1.0 or later, while ensuring that the rest of the user profile is always displayed.
```
